File: moderate/cash_register.c
#include <stdbool.h>
#include <stdio.h>

static const char *units[] = {
	"PENNY", "NICKEL", "DIME", "QUARTER", "HALF DOLLAR", "ONE",
	"TWO", "FIVE", "TEN", "TWENTY", "FIFTY", "ONE HUNDRED"
};
static const int value[] = {
	1, 5, 10, 25, 50, 100,
	200, 500, 1000, 2000, 5000, 10000
};
static const int sorted[] = {
	2, 10, 7, 4, 1, 5, 11, 0, 3, 8, 9, 6
};

int main(int argc, char *argv[]) {
	FILE *fp;
	int p, p1, c, c1, i;

	if (argc != 2) {
		printf("Usage: %s [FILE]\n", argv[0]);
		return 1;
	}
	fp = fopen(*++argv, "r");
	while (fscanf(fp, "%d", &p1) != EOF) {
		int p2 = 0, c2 = 0;
		if (getc(fp) == '.')
			fscanf(fp, "%d%*c", &p2);
		fscanf(fp, "%d", &c1);
		if (getc(fp) == '.')
			fscanf(fp, "%d%*c", &c2);
		p = p1 * 100 + p2;
		c = c1 * 100 + c2;
		if (p > c) {
			printf("ERROR");
		} else if (p == c) {
			printf("ZERO");
		} else {
			int change[12] = { 0 };
			bool con = false;
			for (i = 11; i >= 0; i--) {
				while (c - p >= value[i]) {
					change[i]++;
					c -= value[i];
				}
			}
			for (i = 0; i < 12; i++) {
				while (change[sorted[i]]) {
					if (con) {
						printf(",");
					} else {
						con = true;
					}
					printf("%s", units[sorted[i]]);
					change[sorted[i]]--;
				}
			}
		}
		printf("\n");
	}
	return 0;
}

File: moderate/cash_register.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

var (
	units []string
	value []int
)

func init() {
	units = []string{"PENNY", "NICKEL", "DIME", "QUARTER", "HALF DOLLAR",
		"ONE", "TWO", "FIVE", "TEN", "TWENTY", "FIFTY", "ONE HUNDRED"}
	value = []int{1, 5, 10, 25, 50, 100,
		200, 500, 1000, 2000, 5000, 10000}
}

func cash(q string) string {
	s := strings.Split(q, ";")
	var p, c, v int
	if strings.Contains(s[0], ".") {
		fmt.Sscanf(s[0], "%d.%d", &p, &v)
	} else {
		fmt.Sscan(s[0], &p)
	}
	p, v = p*100+v, 0
	if strings.Contains(s[1], ".") {
		fmt.Sscanf(s[1], "%d.%d", &c, &v)
	} else {
		fmt.Sscan(s[1], &c)
	}
	c = c*100 + v
	if p > c {
		return "ERROR"
	} else if p == c {
		return "ZERO"
	}
	var r []string
	change := make([]int, 12)
	for i := 11; i >= 0; i-- {
		for c-p >= value[i] {
			change[i]++
			c -= value[i]
		}
	}
	for i := 11; i >= 0; i-- {
		for change[i] > 0 {
			r = append(r, units[i])
			change[i]--
		}
	}
	return strings.Join(r, ",")
}

func main() {
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	scanner := bufio.NewScanner(data)
	for scanner.Scan() {
		fmt.Println(cash(scanner.Text()))
	}
}
